File: libmuscle/cpp/src/libmuscle/tests/test_port_manager.cpp
// Inject mocks

// into the real implementation under test.
#include <ymmsl/ymmsl.hpp>

#include <libmuscle/peer_info.cpp>
#include <libmuscle/port.cpp>
#include <libmuscle/port_manager.cpp>


// Test code dependencies
#include <memory>
#include <numeric>
#include <stdexcept>
#include <map>
#include <gtest/gtest.h>
#include <libmuscle/namespace.hpp>
#include <libmuscle/port.hpp>


int main(int argc, char *argv[]) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

using libmuscle::_MUSCLE_IMPL_NS::Optional;
using libmuscle::_MUSCLE_IMPL_NS::PeerDims;
using libmuscle::_MUSCLE_IMPL_NS::PeerInfo;
using libmuscle::_MUSCLE_IMPL_NS::PeerLocations;
using libmuscle::_MUSCLE_IMPL_NS::Port;
using libmuscle::_MUSCLE_IMPL_NS::PortsDescription;
using libmuscle::_MUSCLE_IMPL_NS::PortManager;

using ymmsl::Conduit;
using ymmsl::Identifier;
using ymmsl::Operator;
using ymmsl::Reference;


struct libmuscle_port_manager : ::testing::Test {
    std::vector<int> index_;
    PortManager port_manager_;

    std::vector<int> index2_;
    PortManager port_manager2_;

    libmuscle_port_manager()
        : index_({13})
        , port_manager_(index_, PortsDescription{
                {Operator::O_I, {"out"}},
                {Operator::S, {"in"}}})
        , index2_({})
        , port_manager2_(index2_, PortsDescription{
                {Operator::F_INIT, {"in[]"}},
                {Operator::O_F, {"out[]"}}})
    {
        Reference component_id("other");
        std::vector<Conduit> conduits({
                Conduit("component.out", "other.in"),
                Conduit("other.out", "component.in")});
        PeerDims peer_dims({ {"component", {20}}});
        PeerLocations peer_locations({{"component", {"direct:test"}}});
        PeerInfo peer_info(component_id, index2_, conduits, peer_dims, peer_locations);

        port_manager2_.connect_ports(peer_info);
    }
};


/* Tests */
TEST_F(libmuscle_port_manager, test_connect_ports) {
    Reference component_id("component");
    std::vector<Conduit> conduits({
        Conduit("component.out", "other.in"),
        Conduit("other.settings_out", "component.muscle_settings_in"),
        Conduit("other.out", "component.in")});
    PeerDims peer_dims({{Reference("other"), {}}});
    PeerLocations peer_locations({
            {Reference("other"), {"direct:test"}}});
    PeerInfo peer_info(component_id, index_, conduits, peer_dims, peer_locations);

    port_manager_.connect_ports(peer_info);

    // check automatic ports
    ASSERT_TRUE(port_manager_.settings_in_connected());
    ASSERT_EQ(port_manager_.muscle_settings_in().name, "muscle_settings_in");
    ASSERT_EQ(port_manager_.muscle_settings_in().oper, Operator::F_INIT);
    ASSERT_EQ(port_manager_.muscle_settings_in().length_, -1);

    // check declared ports
    auto const & ports = port_manager_.ports_;

    ASSERT_EQ(ports.at("in").name, "in");
    ASSERT_EQ(ports.at("in").oper, Operator::S);
    ASSERT_EQ(ports.at("in").length_, -1);

    ASSERT_EQ(ports.at("out").name, "out");
    ASSERT_EQ(ports.at("out").oper, Operator::O_I);
    ASSERT_EQ(ports.at("out").length_, -1);
}

TEST_F(libmuscle_port_manager, test_connect_vector_ports) {
    PortsDescription declared_ports({
            {Operator::F_INIT, {"in[]"}},
            {Operator::O_F, {"out1", "out2[]"}}
            });

    PortManager port_manager(index_, declared_ports);

    Reference component_id("component");
    std::vector<Conduit> conduits({
        Conduit("other1.out", "component.in"),
        Conduit("component.out1", "other.in"),
        Conduit("component.out2", "other3.in")
        });
    PeerDims peer_dims({
            {Reference("other1"), {20, 7}},
            {Reference("other"), {25}},
            {Reference("other3"), {20}}
            });
    PeerLocations peer_locations({
            {Reference("other"), {"tcp:test"}},
            {Reference("other1"), {"tcp:test1"}},
            {Reference("other3"), {"tcp:test3"}}
            });
    PeerInfo peer_info(component_id, index_, conduits, peer_dims, peer_locations);

    port_manager.connect_ports(peer_info);

    auto const & ports = port_manager.ports_;

    ASSERT_EQ(ports.at("in").name, "in");
    ASSERT_EQ(ports.at("in").oper, Operator::F_INIT);
    ASSERT_TRUE(ports.at("in").is_vector());
    ASSERT_EQ(ports.at("in").length_, 7);
    ASSERT_FALSE(ports.at("in").is_resizable_);

    ASSERT_EQ(ports.at("out1").name, "out1");
    ASSERT_EQ(ports.at("out1").oper, Operator::O_F);
    ASSERT_EQ(ports.at("out1").length_, -1);

    ASSERT_EQ(ports.at("out2").name, "out2");
    ASSERT_EQ(ports.at("out2").oper, Operator::O_F);
    ASSERT_EQ(ports.at("out2").length_, 0);
    ASSERT_TRUE(ports.at("out2").is_resizable_);
}

TEST_F(libmuscle_port_manager, test_connect_multidimensional_ports) {
    PortsDescription declared_ports({
            {Operator::F_INIT, {"in[][]"}}
            });

    PortManager port_manager(index_, declared_ports);

    Reference component_id("component");
    std::vector<Conduit> conduits({
        Conduit("other.out", "component.in")
        });
    PeerDims peer_dims({
            {Reference("other"), {20, 7, 30}}
            });
    PeerLocations peer_locations({
            {Reference("other"), {"direct:test"}}
            });
    PeerInfo peer_info(component_id, index_, conduits, peer_dims, peer_locations);

    ASSERT_THROW(
            port_manager.connect_ports(peer_info), std::invalid_argument);
}

TEST_F(libmuscle_port_manager, test_connect_inferred_ports) {
    PortManager port_manager(index_, {});

    Reference component_id("component");
    std::vector<Conduit> conduits({
        Conduit("other1.out", "component.in"),
        Conduit("component.out1", "other.in"),
        Conduit("component.out3", "other2.in")
        });
    PeerDims peer_dims({
            {Reference("other1"), {20, 7}},
            {Reference("other"), {25}},
            {Reference("other2"), {}}
            });
    PeerLocations peer_locations({
            {Reference("other"), {"tcp:test"}},
            {Reference("other1"), {"tcp:test1"}},
            {Reference("other2"), {"tcp:test2"}}
            });
    PeerInfo peer_info(component_id, index_, conduits, peer_dims, peer_locations);

    port_manager.connect_ports(peer_info);

    auto const & ports = port_manager.ports_;

    ASSERT_EQ(ports.at("in").name, "in");
    ASSERT_EQ(ports.at("in").oper, Operator::F_INIT);
    ASSERT_EQ(ports.at("in").length_, 7);
    ASSERT_FALSE(ports.at("in").is_resizable_);

    ASSERT_EQ(ports.at("out1").name, "out1");
    ASSERT_EQ(ports.at("out1").oper, Operator::O_F);
    ASSERT_EQ(ports.at("out1").length_, -1);

    ASSERT_EQ(ports.at("out3").name, "out3");
    ASSERT_EQ(ports.at("out3").oper, Operator::O_F);
    ASSERT_EQ(ports.at("out3").length_, -1);
}

TEST_F(libmuscle_port_manager, test_port_message_counts) {
    port_manager_.connect_ports(PeerInfo("component", {}, {}, {}, {}));

    auto msg_counts = port_manager_.get_message_counts();
    ASSERT_EQ(msg_counts.size(), 3);
    ASSERT_EQ(msg_counts["in"], std::vector<int>({0}));
    ASSERT_EQ(msg_counts["out"], std::vector<int>({0}));
    ASSERT_EQ(msg_counts["muscle_settings_in"], std::vector<int>({0}));

    port_manager_.get_port("in").increment_num_messages();

    auto msg_counts2 = port_manager_.get_message_counts();
    ASSERT_EQ(msg_counts2.size(), 3);
    ASSERT_EQ(msg_counts2["in"], std::vector<int>({1}));
    ASSERT_EQ(msg_counts2["out"], std::vector<int>({0}));
    ASSERT_EQ(msg_counts2["muscle_settings_in"], std::vector<int>({0}));

    port_manager_.get_port("out").increment_num_messages();
    port_manager_.get_port("out").increment_num_messages();

    auto msg_counts3 = port_manager_.get_message_counts();
    ASSERT_EQ(msg_counts3.size(), 3);
    ASSERT_EQ(msg_counts3["in"], std::vector<int>({1}));
    ASSERT_EQ(msg_counts3["out"], std::vector<int>({2}));
    ASSERT_EQ(msg_counts3["muscle_settings_in"], std::vector<int>({0}));

    port_manager_.muscle_settings_in().increment_num_messages();

    auto msg_counts4 = port_manager_.get_message_counts();
    ASSERT_EQ(msg_counts4.size(), 3);
    ASSERT_EQ(msg_counts4["in"], std::vector<int>({1}));
    ASSERT_EQ(msg_counts4["out"], std::vector<int>({2}));
    ASSERT_EQ(msg_counts4["muscle_settings_in"], std::vector<int>({1}));

    port_manager_.restore_message_counts(msg_counts);

    auto msg_counts5 = port_manager_.get_message_counts();
    ASSERT_EQ(msg_counts5["in"], std::vector<int>({0}));
    ASSERT_EQ(msg_counts5["out"], std::vector<int>({0}));
    ASSERT_EQ(msg_counts5["muscle_settings_in"], std::vector<int>({0}));

    ASSERT_THROW(
            port_manager_.restore_message_counts({{"x?invalid_port", {3}}}),
            std::runtime_error);
}

TEST_F(libmuscle_port_manager, test_vector_port_message_counts) {
    auto msg_counts = port_manager2_.get_message_counts();
    ASSERT_EQ(msg_counts.size(), 3);
    std::vector<int> expected_counts(20);  // 20 zeros
    ASSERT_EQ(msg_counts["out"], expected_counts);
    ASSERT_EQ(msg_counts["in"], expected_counts);
    ASSERT_EQ(msg_counts["muscle_settings_in"], std::vector<int>({0}));

    port_manager2_.get_port("out").increment_num_messages(13);

    msg_counts = port_manager2_.get_message_counts();
    ASSERT_EQ(msg_counts.size(), 3);
    ASSERT_EQ(msg_counts["in"], expected_counts);
    ASSERT_EQ(msg_counts["muscle_settings_in"], std::vector<int>({0}));
    expected_counts[13] = 1;
    ASSERT_EQ(msg_counts["out"], expected_counts);

    std::iota(expected_counts.begin(), expected_counts.end(), 0);
    port_manager2_.restore_message_counts({
            {"out", expected_counts},
            {"in", expected_counts},
            {"muscle_settings_in", {4}}});

    port_manager2_.get_port("out").increment_num_messages(13);

    msg_counts = port_manager2_.get_message_counts();
    ASSERT_EQ(msg_counts.size(), 3);
    ASSERT_EQ(msg_counts["in"], expected_counts);
    ASSERT_EQ(msg_counts["muscle_settings_in"], std::vector<int>({4}));
    expected_counts[13] = 14;
    ASSERT_EQ(msg_counts["out"], expected_counts);
}

File: libmuscle/python/libmuscle/test/test_port_manager.py
from typing import List
from libmuscle.peer_info import PeerInfo
from libmuscle.port_manager import PortManager

from ymmsl import Conduit, Identifier, Operator, Reference as Ref

import pytest


@pytest.fixture
def index() -> List[int]:
    return [13]


@pytest.fixture
def port_manager(index) -> PortManager:
    declared_ports = {
            Operator.O_I: ['out'],
            Operator.S: ['in']}
    return PortManager(index, declared_ports)


@pytest.fixture
def index2() -> List[int]:
    return []


@pytest.fixture
def port_manager2(index2) -> PortManager:
    declared_ports = {
            Operator.F_INIT: ['in[]'],
            Operator.O_F: ['out[]']}
    port_manager = PortManager(index2, declared_ports)

    component_id = Ref('other')
    conduits = [Conduit('component.out', 'other.in'),
                Conduit('other.out', 'component.in')]
    peer_dims = {Ref('component'): [20]}
    peer_locations = {Ref('component'): ['direct:test']}
    peer_info = PeerInfo(component_id, index2, conduits, peer_dims, peer_locations)

    port_manager.connect_ports(peer_info)
    return port_manager


def test_connect_ports(index, port_manager) -> None:
    component_id = Ref('component')
    conduits = [Conduit('component.out', 'other.in'),
                Conduit('other.settings_out', 'component.muscle_settings_in'),
                Conduit('other.out', 'component.in')]
    peer_dims = {Ref('other'): []}
    peer_locations = {Ref('other'): ['direct:test']}
    peer_info = PeerInfo(component_id, index, conduits, peer_dims, peer_locations)

    port_manager.connect_ports(peer_info)

    # check automatic ports
    assert port_manager.settings_in_connected()
    assert port_manager._muscle_settings_in.name == Identifier('muscle_settings_in')
    assert port_manager._muscle_settings_in.operator == Operator.F_INIT
    assert port_manager._muscle_settings_in._length is None

    # check declared ports
    ports = port_manager._ports
    assert ports['in'].name == Identifier('in')
    assert ports['in'].operator == Operator.S
    assert ports['in']._length is None

    assert ports['out'].name == Identifier('out')
    assert ports['out'].operator == Operator.O_I
    assert ports['out']._length is None


def test_connect_vector_ports(index) -> None:
    declared_ports = {
            Operator.F_INIT: ['in[]'],
            Operator.O_F: ['out1', 'out2[]']}

    port_manager = PortManager(index, declared_ports)

    component_id = Ref('component')
    conduits = [Conduit('other1.out', 'component.in'),
                Conduit('component.out1', 'other.in'),
                Conduit('component.out2', 'other3.in')]
    peer_dims = {
            Ref('other1'): [20, 7],
            Ref('other'): [25],
            Ref('other3'): [20]}
    peer_locations = {
            Ref('other'): ['direct:test'],
            Ref('other1'): ['direct:test1'],
            Ref('other3'): ['direct:test3']}
    peer_info = PeerInfo(component_id, index, conduits, peer_dims, peer_locations)

    port_manager.connect_ports(peer_info)

    ports = port_manager._ports
    assert ports['in'].name == Identifier('in')
    assert ports['in'].operator == Operator.F_INIT
    assert ports['in']._length == 7
    assert ports['in']._is_resizable is False

    assert ports['out1'].name == Identifier('out1')
    assert ports['out1'].operator == Operator.O_F
    assert ports['out1']._length is None

    assert ports['out2'].name == Identifier('out2')
    assert ports['out2'].operator == Operator.O_F
    assert ports['out2']._length == 0
    assert ports['out2']._is_resizable is True


def test_connect_multidimensional_ports() -> None:
    index = [13]
    declared_ports = {Operator.F_INIT: ['in[][]']}
    port_manager = PortManager(index, declared_ports)

    component_id = Ref('component')
    conduits = [Conduit(Ref('other.out'), Ref('component.in'))]
    peer_dims = {Ref('other'): [20, 7, 30]}
    peer_locations = {Ref('other'): ['direct:test']}
    peer_info = PeerInfo(component_id, index, conduits, peer_dims, peer_locations)

    with pytest.raises(ValueError):
        port_manager.connect_ports(peer_info)


def test_connect_inferred_ports() -> None:
    index = [13]
    port_manager = PortManager(index, None)

    component_id = Ref('component')
    conduits = [Conduit('other1.out', 'component.in'),
                Conduit('component.out1', 'other.in'),
                Conduit('component.out3', 'other2.in')]
    peer_dims = {
            Ref('other1'): [20, 7],
            Ref('other'): [25],
            Ref('other2'): []}
    peer_locations = {
            Ref('other'): ['direct:test'],
            Ref('other1'): ['direct:test1'],
            Ref('other2'): ['direct:test2']}
    peer_info = PeerInfo(component_id, index, conduits, peer_dims, peer_locations)

    port_manager.connect_ports(peer_info)

    ports = port_manager._ports
    assert ports['in'].name == Identifier('in')
    assert ports['in'].operator == Operator.F_INIT
    assert ports['in']._length == 7
    assert ports['in']._is_resizable is False

    assert ports['out1'].name == Identifier('out1')
    assert ports['out1'].operator == Operator.O_F
    assert ports['out1']._length is None

    assert ports['out3'].name == Identifier('out3')
    assert ports['out3'].operator == Operator.O_F
    assert ports['out3']._length is None


def test_port_message_counts(port_manager) -> None:
    port_manager.connect_ports(PeerInfo('component', [], [], {}, {}))

    msg_counts = port_manager.get_message_counts()
    assert msg_counts == {'in': [0], 'out': [0], 'muscle_settings_in': [0]}

    port_manager.get_port('in').increment_num_messages()

    msg_counts2 = port_manager.get_message_counts()
    assert msg_counts2 == {'in': [1], 'out': [0], 'muscle_settings_in': [0]}

    port_manager.get_port('out').increment_num_messages()
    port_manager.get_port('out').increment_num_messages()

    msg_counts3 = port_manager.get_message_counts()
    assert msg_counts3 == {'in': [1], 'out': [2], 'muscle_settings_in': [0]}

    port_manager._muscle_settings_in.increment_num_messages()

    msg_counts4 = port_manager.get_message_counts()
    assert msg_counts4 == {'in': [1], 'out': [2], 'muscle_settings_in': [1]}

    port_manager.restore_message_counts(msg_counts)

    msg_counts5 = port_manager.get_message_counts()
    assert msg_counts5 == {'in': [0], 'out': [0], 'muscle_settings_in': [0]}

    with pytest.raises(RuntimeError):
        port_manager.restore_message_counts({"x?invalid_port": 3})


def test_vector_port_message_counts(port_manager2) -> None:
    msg_counts = port_manager2.get_message_counts()
    assert msg_counts == {'out': [0] * 20,
                          'in': [0] * 20,
                          'muscle_settings_in': [0]}

    port_manager2.get_port('out').increment_num_messages(13)

    msg_counts = port_manager2.get_message_counts()
    assert msg_counts == {'out': [0] * 13 + [1] + [0] * 6,
                          'in': [0] * 20,
                          'muscle_settings_in': [0]}

    port_manager2.restore_message_counts({'out': list(range(20)),
                                          'in': list(range(20)),
                                          'muscle_settings_in': [4]})

    port_manager2.get_port('out').increment_num_messages(13)

    msg_counts = port_manager2.get_message_counts()
    assert msg_counts == {'out': list(range(13)) + [14] + list(range(14, 20)),
                          'in': list(range(20)),
                          'muscle_settings_in': [4]}
